pkg/event/target/queuestore.go

```diff
@@ -0,0 +1,219 @@
+/*
+ * Minio Cloud Storage, (C) 2018 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package target
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/minio/minio/pkg/event"
+)
+
+const (
+	maxLimit = 10000 // Max store limit.
+	eventExt = ".event"
+)
+
+type QueueStore struct {
+	sync.RWMutex
+	directory string
+	eC        uint16
+	limit     uint16
+}
+
+// NewQueueStore - Filestore for persisting events.
+func NewQueueStore(directory string, limit uint16) *QueueStore {
+	if limit == 0 {
+		limit = maxLimit
+	}
+	queueStore := &QueueStore{
+		directory: directory,
+		limit:     limit,
+	}
+	return queueStore
+}
+
+// Staging directory.
+func tmpPath(directory string) string {
+	return directory + "/tmp"
+}
+
+// Open - Creates the directory if not present.
+func (store *QueueStore) Open() error {
+	store.Lock()
+	defer store.Unlock()
+
+	if err := store.resetTmp(); err != nil {
+		return err
+	}
+
+	tmpDir := tmpPath(store.directory)
+	if terr := os.MkdirAll(tmpDir, os.FileMode(0770)); terr != nil {
+		return terr
+	}
+
+	eCount := uint16(len(store.listAll()))
+	if eCount >= store.limit {
+		return ErrLimitExceeded
+	} else {
+		store.eC = eCount
+	}
+
+	return nil
+}
+
+// write - writes event to the directory.
+func (store *QueueStore) write(directory string, key string, e event.Event) error {
+
+	tp := filepath.Join(tmpPath(store.directory), key)
+
+	// Creates the temporary file.
+	temp, cerr := os.Create(tp)
+	if cerr != nil {
+		return cerr
+	}
+
+	// Marshalls the event.
+	eventData, err := json.Marshal(e)
+	if err != nil {
+		return err
+	}
+
+	// Writes to the temporary file.
+	if _, err := temp.Write(eventData); err != nil {
+		return err
+	}
+
+	p := filepath.Join(store.directory, key+eventExt)
+	// Rename the temp file to the destination file.
+	if err := os.Rename(tp, p); err != nil {
```

Comment:
This type of `/tmp` optimization and protection is perhaps not needed, these files are really not worth saving if the server is about to crash. Since partial data that is saved at `/tmp` would require another additional cleanup. 

The best way is just to use `ioutil.WriteFile()` and be done with it.

Reply:
:+1: